File: php/src/Providers/ZoromailProvider.php
<?php

declare(strict_types=1);

namespace TempMail\Providers;

use TempMail\HttpClient;
use TempMail\Models\Message;
use TempMail\Models\MessageDetail;
use TempMail\TempMailProvider;
use TempMail\Exceptions\TempMailException;

class ZoromailProvider implements TempMailProvider
{
    private const BASE_URL = 'https://zoromail.com/public_api.php/v1';

    private HttpClient $http;
    private ?string $currentEmail = null;

    public function __construct(?HttpClient $http = null)
    {
        $this->http = $http ?? new HttpClient();
    }

    /** Call zoromail API; checks {success, data, error} envelope. */
    private function api(string $method, string $path, ?array $body = null): mixed
    {
        $res = $method === 'GET'
            ? $this->http->get(self::BASE_URL . $path)
            : $this->http->post(self::BASE_URL . $path, $body ?? []);

        $data = $res['body'];
        if (!is_array($data) || !($data['success'] ?? false)) {
            throw new TempMailException('zoromail: ' . (is_array($data) ? ($data['error'] ?? 'unknown error') : 'unexpected response'));
        }
        return $data['data'] ?? null;
    }

    public function generateEmail(): string
    {
        $domains = $this->api('GET', '/domains');
        if (!is_array($domains) || empty($domains)) {
            throw new TempMailException('zoromail: no domains available');
        }
        $domain = $domains[array_rand($domains)];
        $username = bin2hex(random_bytes(5)); // 10 chars
        $data = $this->api('POST', '/emails', ['username' => $username, 'domain' => $domain]);
        if (!is_array($data) || !isset($data['email'])) {
            throw new TempMailException('zoromail: no email in response');
        }
        $this->currentEmail = $data['email'];
        return $this->currentEmail;
    }

    public function getInbox(string $email): array
    {
        $data = $this->api('GET', '/emails/' . urlencode($email) . '/messages');
        $messages = [];
        foreach ($data ?? [] as $item) {
            $messages[] = new Message(
                id: (string)($item['id'] ?? ''),
                sender: $item['from'] ?? '',
                subject: $item['subject'] ?? '',
                date: $this->parseDate($item['date'] ?? null),
            );
        }
        return $messages;
    }

    public function readMessage(string $email, string $messageId): MessageDetail
    {
        $data = $this->api('GET', '/messages/' . urlencode($messageId));
        if (!is_array($data)) {
            throw new TempMailException('zoromail: invalid message response');
        }
        return new MessageDetail(
            id: (string)($data['id'] ?? $messageId),
            sender: $data['from'] ?? '',
            subject: $data['subject'] ?? '',
            date: $this->parseDate($data['date'] ?? null),
            bodyText: $data['text'] ?? $data['body_text'] ?? '',
            bodyHtml: $data['html'] ?? $data['body_html'] ?? '',
        );
    }

    public function deleteEmail(string $email): bool
    {
        $this->currentEmail = null;
        return true;
    }

    public function waitForEmail(string $email, int $timeout = 60, int $interval = 5): ?Message
    {
        $deadline = time() + $timeout;
        while (time() < $deadline) {
            if (!empty($inbox = $this->getInbox($email))) {
                return $inbox[0];
            }
            sleep($interval);
        }
        return null;
    }

    private function parseDate(?string $str): \DateTimeImmutable
    {
        if ($str !== null && $str !== '') {
            try {
                return new \DateTimeImmutable($str);
            } catch (\Throwable) {}
        }
        return new \DateTimeImmutable();
    }
}

File: java/src/main/java/com/tempmail/providers/ZoromailProvider.java
package com.tempmail.providers;

import com.tempmail.TempMailProvider;
import com.tempmail.errors.TempMailException;
import com.tempmail.http.HttpClient;
import com.tempmail.json.SimpleJson;
import com.tempmail.json.SimpleJson.JsonArray;
import com.tempmail.json.SimpleJson.JsonObject;
import com.tempmail.json.SimpleJson.JsonValue;
import com.tempmail.models.Message;
import com.tempmail.models.MessageDetail;

import java.time.Duration;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Random;

/**
 * Zoromail provider - REST API, no auth. Response envelope: {success, data, error}.
 */
public class ZoromailProvider implements TempMailProvider {

    private static final String BASE_URL = "https://zoromail.com/public_api.php/v1";
    private static final String CHARS = "abcdefghijklmnopqrstuvwxyz0123456789";
    private static final DateTimeFormatter DATE_FMT = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");

    private final HttpClient client;
    private final Random random = new Random();
    private String email;

    public ZoromailProvider() {
        this.client = new HttpClient(true, false);
    }

    @Override
    public String generateEmail() throws TempMailException {
        JsonArray domains = api("GET", "/domains", null).asArray();
        if (domains == null || domains.isEmpty()) throw new TempMailException("zoromail: no domains available");
        JsonValue d = domains.get(random.nextInt(domains.size()));
        String domain = d.isObject() ? d.asObject().getString("domain") : d.asString();
        String username = randUser();
        JsonObject body = SimpleJson.object();
        body.put("username", SimpleJson.val(username));
        body.put("domain", SimpleJson.val(domain));
        JsonObject data = api("POST", "/emails", body.toJson()).asObject();
        email = data.getString("email");
        if (email == null || email.isEmpty()) throw new TempMailException("zoromail: no email in response");
        return email;
    }

    @Override
    public List<Message> getInbox(String email) throws TempMailException {
        JsonArray items = api("GET", "/emails/" + email + "/messages", null).asArray();
        List<Message> messages = new ArrayList<>();
        if (items == null) return messages;
        for (JsonValue el : items.elements()) {
            JsonObject item = el.asObject();
            messages.add(new Message(
                    item.getString("id"),
                    item.getString("from"),
                    item.getString("subject"),
                    parseDate(item.getString("date"))
            ));
        }
        return messages;
    }

    @Override
    public MessageDetail readMessage(String messageId) throws TempMailException {
        JsonObject data = api("GET", "/messages/" + messageId, null).asObject();
        String text = data.getString("text");
        String html = data.getString("html");
        return new MessageDetail(
                data.getString("id"),
                data.getString("from"),
                data.getString("subject"),
                parseDate(data.getString("date")),
                text != null ? text : "",
                html != null ? html : "",
                Collections.emptyList()
        );
    }

    @Override
    public boolean deleteEmail(String email) throws TempMailException {
        this.email = null;
        return true;
    }

    @Override
    public Optional<Message> waitForEmail(String email, Duration timeout, Duration interval) throws TempMailException {
        long deadline = System.currentTimeMillis() + timeout.toMillis();
        while (System.currentTimeMillis() < deadline) {
            List<Message> inbox = getInbox(email);
            if (!inbox.isEmpty()) return Optional.of(inbox.get(0));
            try {
                Thread.sleep(interval.toMillis());
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                throw new TempMailException("Interrupted while waiting for email", e);
            }
        }
        return Optional.empty();
    }

    // Returns the "data" field after checking success.
    private JsonValue api(String method, String path, String body) throws TempMailException {
        String resp = "GET".equals(method)
                ? client.get(BASE_URL + path)
                : client.post(BASE_URL + path, body, Map.of("Content-Type", "application/json"));
        JsonObject root = SimpleJson.parseObject(resp);
        if (!root.getBoolean("success")) {
            throw new TempMailException("zoromail API error: " + root.getString("error"));
        }
        return root.get("data");
    }

    private String randUser() {
        StringBuilder sb = new StringBuilder(10);
        for (int i = 0; i < 10; i++) sb.append(CHARS.charAt(random.nextInt(CHARS.length())));
        return sb.toString();
    }

    private LocalDateTime parseDate(String s) {
        if (s == null || s.isEmpty()) return LocalDateTime.now();
        try {
            String t = s.replace("Z", "").replace("+00:00", "");
            if (t.contains("T")) return LocalDateTime.parse(t);
            return LocalDateTime.parse(t, DATE_FMT);
        } catch (Exception e) {
            return LocalDateTime.now();
        }
    }
}
